esrally/utils/net.py

```diff
@@ -94,17 +95,64 @@ def __call__(self, bytes_amount):
                          Config=boto3.s3.transfer.TransferConfig(use_threads=False))
 
 
-def download_s3(url, local_path, expected_size_in_bytes=None, progress_indicator=None):
+def _build_gcs_object_url(bucket_name, bucket_path):
+    # / and other special characters must be urlencoded in bucket and object names
+    # ref: https://cloud.google.com/storage/docs/request-endpoints#encoding
+
+    return functools.reduce(urllib.parse.urljoin, [
+        "https://storage.googleapis.com/storage/v1/b/",
+        f"{quote(bucket_name.strip('/'), safe='')}/",
+        "o/",
+        f"{quote(bucket_path.strip('/'), safe='')}",
+        "?alt=media"
+    ])
+
+
+def _download_from_gcs_bucket(bucket_name, bucket_path, local_path, expected_size_in_bytes=None, progress_indicator=None):
+    # lazily initialize Google Cloud Storage support - we might not need it
+    import google.oauth2.credentials
+    import google.auth.transport.requests as tr_requests
+    import google.auth
+    # Using Google Resumable Media as the standard storage library doesn't support progress
+    # (https://github.com/googleapis/python-storage/issues/27)
+    from google.resumable_media.requests import ChunkedDownload
+    ro_scope = "https://www.googleapis.com/auth/devstorage.read_only"
+
+    access_token = os.environ.get("GOOGLE_AUTH_TOKEN")
+    if access_token:
+        credentials = google.oauth2.credentials.Credentials(token=access_token, scopes=(ro_scope, ))
+    else:
+        # https://google-auth.readthedocs.io/en/latest/user-guide.html
+        credentials, _ = google.auth.default(scopes=(ro_scope,))
+
+    transport = tr_requests.AuthorizedSession(credentials)
+    chunk_size = 50 * 1024 * 1024  # 50MB
+
+    with open(local_path, "wb") as local_fp:
+        media_url = _build_gcs_object_url(bucket_name, bucket_path)
+        download = ChunkedDownload(media_url, chunk_size, local_fp)
+        # allow us to calculate the total bytes
+        download.consume_next_chunk(transport)
+        if not expected_size_in_bytes:
+            expected_size_in_bytes = download.total_bytes
+        while not download.finished:
+            if progress_indicator and download.bytes_downloaded and download.total_bytes:
```

Comment:
is there ever a case where we wont have a `progress_indicator`?

Reply:
We use it in https://github.com/elastic/rally/blob/1bdec4ff774e9c66640124437189bda1dd9cc58d/esrally/mechanic/supplier.py#L519 and https://github.com/elastic/rally/blob/1bdec4ff774e9c66640124437189bda1dd9cc58d/esrally/track/loader.py#L410 but want to be able to support a case (in the future?) were it shouldn't be used.